### PR TITLE
Transaction background indexing/unindexing + improved chain initiation from freezer

### PR DESCRIPTION
**Pull request by holiman**
This is an (as yet pretty untested) PR which is an alternative to https://github.com/ethereum/go-ethereum/pull/20270 . 

It implements a few changes, 

- The hashing of transactions are based on hashing the raw rlp-data delivered from the database(s). As such, it skips the overhead of unmarshalling an object and then marshalling it again in order to hash it. 
- The lookup is improved. 
  - For data that is in ancients (canonical only), we can request it by number _only_, and do not need to lookup hash, header, block body, but can go directly to body. 
  - For data that is not yet in ancients, it is also an improvement, we need only obtain the hash, then read body rlp.
- The initiation of a chain from ancients, where we only require writing the hash to number mappings, has been heavily improved. This PR reads hashes directly from ancients and shoves them into leveldb without any other overhead or lookups. 
- The new "generic" iterator for transaction hashes does not _guarantee_ strict ordering. 

I have not yet tested it, and I'm not 100% that I maintained the original semantics of `from`, `to` in regards to the iteration over a span of blocks. 

The `iterateCanonicalChain` is (in this PR) totally unused aside from tests, and should be removed. I left it in there for now, I need to rewrite the existing tests (which are the only things that make use of it) to use the new code instead (EDIT: done now)

Also, the RLP iteration will likely change.

### CODE REVIEW

**Comment by rjl493456442** [comment]
> The new "generic" iterator for transaction hashes does not guarantee strict ordering.

I guess it's easy to implement. For the `delivery` channel, we can use a `priorityQueue` and sort them by block number.

**Comment by holiman** [comment]
> For the delivery channel, we can use a priorityQueue and sort them by block number.

Yep, that makes sense

**Comment by holiman** [comment]
This has now been fixed, PTAL

File: core/rawdb/chain_iterator.go [hunk 1]
```diff
@@ -28,157 +28,164 @@ import (
...
+		// If enough data was accumulated in memory or we're at the last block, dump to disk
+		if batch.ValueSize() > ethdb.IdealBatchSize {
+			if err := batch.Write(); err != nil {
+				log.Crit("Failed to write data to db", "err", err)
+			}
+			batch.Reset()
+		}
+		// If we've spent too much time already, notify the user of what we're doing
+		if time.Since(logged) > 8*time.Second {
+			log.Info("Initializing database from freezer", "blocks", i, "total", frozen, "tail", i, "hash", hash, "elapsed", common.PrettyDuration(time.Since(start)))
```
Comment (rjl493456442) [inline core/rawdb/chain_iterator.go]:
```
log.Info("Initializing database from freezer", "total", frozen, "number", i, "hash", hash, "elapsed", common.PrettyDuration(time.Since(start)))
```

File: core/rawdb/chain_iterator.go [hunk 2]
```diff
@@ -28,157 +28,164 @@ import (
...
+		log.Crit("Failed to write data to db", "err", err)
+	}
+	batch.Reset()
 
-	threads := to - from
-	if cpus := runtime.NumCPU(); threads > uint64(cpus) {
-		threads = uint64(cpus)
+	WriteHeadHeaderHash(db, hash)
+	WriteHeadFastBlockHash(db, hash)
+	log.Info("Initialized database from freezer", "blocks", frozen, "tail", frozen, "elapsed", common.PrettyDuration(time.Since(start)))
```
Comment (rjl493456442) [inline core/rawdb/chain_iterator.go]: Please remove `tail` in log

File: core/rawdb/chain_iterator.go [hunk 3]
```diff
@@ -28,157 +28,164 @@ import (
...
-			return errors.New("broken database")
-		}
-		// Push the block into the import queue and process contiguous ranges
-		priority := -int64(block.NumberU64())
+	// lookup runs in one instance
+	lookup := func() {
+		n, end := from, to
 		if reverse {
-			priority = int64(block.NumberU64())
+			n, end = to-1, from-1
```
Comment (rjl493456442) [inline core/rawdb/chain_iterator.go]: `from` can be 0 here. So potential underflow

File: core/rawdb/chain_iterator.go [hunk 4]
```diff
@@ -28,157 +28,164 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethdb"
 	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"golang.org/x/crypto/sha3"
```
Comment (rjl493456442) [inline core/rawdb/chain_iterator.go]: Why use `golang.org/x/crypto/sha3` here?

File: core/rawdb/chain_iterator.go [hunk 5]
```diff
@@ -187,41 +194,109 @@ func InitDatabaseFromFreezer(db ethdb.Database) {
...
-	writeIndices := func(batch ethdb.Batch, block *types.Block) {
-		WriteTxLookupEntries(batch, block)
-		if block.NumberU64() == to-1 || block.NumberU64()%10000 == 0 {
-			WriteTxIndexTail(batch, block.NumberU64())
+	var (
+		hashesCh, abortCh = iterateTransactions(db, from, to, true)
+		batch             = db.NewBatch()
+		start             = time.Now()
+		logged            = start.Add(-7 * time.Second)
+		lastNum           uint64
```
Comment (rjl493456442) [inline core/rawdb/chain_iterator.go]: Shouldn't `lastNum` be initialized?

File: core/rawdb/chain_iterator.go (hunk 4 above)
Comment (holiman) [reply core/rawdb/chain_iterator.go]: The hasher-processors re-use their hashers, which are `sha3.NewLegacyKeccak256`

File: core/rawdb/chain_iterator.go (hunk 3 above)
Comment (holiman) [reply core/rawdb/chain_iterator.go]: Heh, I added a testcase to catch that, and the thing is, yes, it oveflows.. but so does `n--` further down, so the equality check `for n != end` still works as it ought to

File: core/rawdb/chain_iterator.go (hunk 5 above)
Comment (holiman) [reply core/rawdb/chain_iterator.go]: Not necessarily. If we don't write anything, it will be zero. However, you are right that if that _does_ happen (the chans are aborted for whatever reason), we shouldn't write an index tail of `0` to db..

Comment (holiman) [reply core/rawdb/chain_iterator.go]: maybe
```
lastNum           = *ReadTxIndexTail(db)
```
?

**Comment by holiman** [comment]
Init from freezer was pretty fast: 
```
mon06.ethdevops.io INFO [11-18|13:40:40.153] Initializing database from freezer total=8866792 number=169866 hash=4aa5eb…b58dd7 elapsed=1.000s
mon06.ethdevops.io INFO [11-18|13:40:48.154] Initializing database from freezer total=8866792 number=1343910 hash=46a72c…10feca elapsed=9.000s
mon06.ethdevops.io INFO [11-18|13:40:56.187] Initializing database from freezer total=8866792 number=2432189 hash=8597a0…2c7187 elapsed=17.033s
mon06.ethdevops.io INFO [11-18|13:41:04.187] Initializing database from freezer total=8866792 number=3479752 hash=a126c4…ab661b elapsed=25.033s
mon06.ethdevops.io INFO [11-18|13:41:12.187] Initializing database from freezer total=8866792 number=4497964 hash=e71696…1eb1a2 elapsed=33.033s
mon06.ethdevops.io INFO [11-18|13:41:20.187] Initializing database from freezer total=8866792 number=5483211 hash=5b52c9…c4cd8a elapsed=41.033s
mon06.ethdevops.io INFO [11-18|13:41:28.240] Initializing database from freezer total=8866792 number=6477305 hash=6351e4…030de1 elapsed=49.086s
mon06.ethdevops.io INFO [11-18|13:41:36.240] Initializing database from freezer total=8866792 number=7458883 hash=81c548…ef7d53 elapsed=57.086s
mon06.ethdevops.io INFO [11-18|13:41:44.267] Initializing database from freezer total=8866792 number=8397455 hash=68763e…ed204f elapsed=1m5.113s
mon06.ethdevops.io INFO [11-18|13:41:48.229] Initialized database from freezer blocks=8866792 elapsed=1m9.075s 
```

... but something's still not right: 
```
mon06.ethdevops.io panic: runtime error: invalid memory address or nil pointer dereference
mon06.ethdevops.io [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x907640]
mon06.ethdevops.io goroutine 134 [running]:
mon06.ethdevops.io github.com/ethereum/go-ethereum/core/rawdb.IndexTransactions(0x142ddc0, 0xc00021a120, 0x0, 0x874be8)
mon06.ethdevops.io /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/rawdb/chain_iterator.go:200 +0x170
mon06.ethdevops.io github.com/ethereum/go-ethereum/core.(*BlockChain).maintainTxIndex(0xc020c4c000, 0x874be8)
mon06.ethdevops.io /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/blockchain.go:2124 +0x83
mon06.ethdevops.io created by github.com/ethereum/go-ethereum/core.NewBlockChain
mon06.ethdevops.io /go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/blockchain.go:314 +0xc92 
```

**Comment by holiman** [comment]
Fixed the bug. Now indexing : 
https://my.papertrailapp.com/systems/3718933602/events?q=Indexing%20transactions&selected=1133078385795465263

That's somewhere around 1250 blocks per second -- not sure if we have any figures to compare with? 

todo: add better stat printouts

```
mon06.ethdevops.io INFO [11-18|14:31:46.160] Indexing transactions blocks=8865670 total=8866792 elapsed=1.051s
mon06.ethdevops.io INFO [11-18|14:31:54.390] Indexing transactions blocks=8853354 total=8866792 elapsed=9.281s
mon06.ethdevops.io INFO [11-18|14:32:02.471] Indexing transactions blocks=8844464 total=8866792 elapsed=17.362s
mon06.ethdevops.io INFO [11-18|14:32:10.774] Indexing transactions blocks=8830125 total=8866792 elapsed=25.665s
mon06.ethdevops.io INFO [11-18|14:32:18.945] Indexing transactions blocks=8821527 total=8866792 elapsed=33.836s
mon06.ethdevops.io INFO [11-18|14:32:27.350] Indexing transactions blocks=8812734 total=8866792 elapsed=42.240s
mon06.ethdevops.io INFO [11-18|14:32:35.488] Indexing transactions blocks=8802870 total=8866792 elapsed=50.379s
mon06.ethdevops.io INFO [11-18|14:32:43.653] Indexing transactions blocks=8790340 total=8866792 elapsed=58.544s 
```

**Comment by holiman** [comment]
with a bit better stats:
```
mon06.ethdevops.io INFO [11-18|14:47:07.986] Indexing transactions blocks=1217 txs=171541 total=8877403 elapsed=1.191s
mon06.ethdevops.io INFO [11-18|14:47:16.243] Indexing transactions blocks=11361 txs=1232001 total=8877403 elapsed=9.448s
mon06.ethdevops.io INFO [11-18|14:47:24.289] Indexing transactions blocks=21428 txs=2223696 total=8877403 elapsed=17.494s
mon06.ethdevops.io INFO [11-18|14:47:32.301] Indexing transactions blocks=34562 txs=3728162 total=8877403 elapsed=25.506s
mon06.ethdevops.io INFO [11-18|14:47:40.459] Indexing transactions blocks=42753 txs=4617889 total=8877403 elapsed=33.664s
mon06.ethdevops.io INFO [11-18|14:47:48.726] Indexing transactions blocks=50620 txs=5541668 total=8877403 elapsed=41.931s
mon06.ethdevops.io INFO [11-18|14:47:56.859] Indexing transactions blocks=61188 txs=6670430 total=8877403 elapsed=50.064s
mon06.ethdevops.io INFO [11-18|14:48:04.921] Indexing transactions blocks=71718 txs=7970741 total=8877403 elapsed=58.127s
mon06.ethdevops.io INFO [11-18|14:48:13.051] Indexing transactions blocks=82562 txs=9236684 total=8877403 elapsed=1m6.256s
mon06.ethdevops.io INFO [11-18|14:48:21.084] Indexing transactions blocks=94306 txs=10503515 total=8877403 elapsed=1m14.289s
mon06.ethdevops.io INFO [11-18|14:48:29.233] Indexing transactions blocks=106178 txs=11701616 total=8877403 elapsed=1m22.438s
mon06.ethdevops.io INFO [11-18|14:48:37.440] Indexing transactions blocks=117346 txs=12932871 total=8877403 elapsed=1m30.646s
mon06.ethdevops.io INFO [11-18|14:48:45.504] Indexing transactions blocks=128184 txs=14199573 total=8877403 elapsed=1m38.709s
mon06.ethdevops.io INFO [11-18|14:48:53.520] Indexing transactions blocks=140308 txs=15465088 total=8877403 elapsed=1m46.726s
mon06.ethdevops.io INFO [11-18|14:49:01.616] Indexing transactions blocks=153258 txs=16696087 total=8877403 elapsed=1m54.822s
mon06.ethdevops.io INFO [11-18|14:49:09.737] Indexing transactions blocks=165492 txs=18030721 total=8877403 elapsed=2m2.942s 
```

So `18030721` txs in 2 minutes, that's almost 150K transactions per second, or `1350` blocks per second. If so, I guess it will be done in about 1-1.5h (assuming it will be quicker on some older chain segments)

**Comment by holiman** [comment]
```
 INFO [11-18|15:01:26.729] Indexing transactions blocks=907635 txs=101437851 total=8877403 elapsed=14m19.934s 
```
So after 14 minutes, (while it's also fetching states from network), it's crunched through `118K` txs per second, or `1056` blocks per second.

**Comment by holiman** [comment]
Memory shot up through the roof, though (last spike is the current run)
Screenshot_2019-11-18 Dual Geth - Grafana

**Comment by holiman** [comment]
Ok, with mem spke fixed, 
```
 INFO [11-18|15:30:23.457] Indexing transactions blocks= txs=155865202 current=7511699 total=8880128 elapsed=17m23.272s 
```
`1312` blocks/s , `149K` txs per second. 
Memory: previous run and this run: 
Screenshot_2019-11-18 Dual Geth - Grafana(1)

**Comment by holiman** [comment]
Done in `1h6m`: 
```
INFO [11-18|16:19:43.211] Indexed transactions blocks=8880128 txs=576222797 elapsed=1h6m43.026s 
```
Averaged `2218` blocks per second, `144K` transactions per second

**Comment by holiman** [comment]
I did a quick test with this PR with https://github.com/ethereum/go-ethereum/pull/20308 on top, used during the chain init from ancients. Both PRs work, but there was no real noticeable difference: 
```
Initialized database from freezer blocks=8880128 elapsed=1m0.719s
```
So `60s` instead of `69s`. 

Anyway, from my perspective, this PR is pretty done

**Comment by holiman** [comment]
rebased after conflicts appeared in https://github.com/ethereum/go-ethereum/pull/20403 , also lightly squashed

**Comment by holiman** [comment]
We just did a reinit-from-ancient on one of the bootnodes, _just the blocks_ : 
```
3h7m25.329s
```
So this PR brings 3+ hours down to about a minute

**Comment by holiman** [comment]
Rebased on master

**Comment by holiman** [comment]
From the test-run that @karalabe did :
Step one (block hash index):
```
 Feb 17 12:06:51 mon08.ethdevops.io geth INFO [02-17|11:06:50.995] Initialized database from freezer blocks=9410225 elapsed=1m5.096s 
```
Step two (tx index in background):
```
 Feb 17 13:25:23 mon08.ethdevops.io geth INFO [02-17|12:25:23.347] Indexed transactions blocks=9410225 txs=630852737 elapsed=1h18m32.350s 
```


Mon09 (both indexes, foreground, before sync starts) :
```
 Feb 17 13:19:39 mon09.ethdevops.io geth INFO [02-17|12:19:39.247] Initialized chain from ancient data number=9410224 hash=e84ba7…51ecc0 elapsed=1h13m45.566s 
```

TLDR; `mon08` could start the sync process one hour before `mon09`, and finished all indexing roughly at the time `mon09` started it's sync.

File: cmd/utils/flags.go [hunk 6]
```diff
@@ -225,6 +225,11 @@ var (
 		Usage: `Blockchain garbage collection mode ("full", "archive")`,
 		Value: "full",
 	}
+	TxLookupLimitFlag = cli.Int64Flag{
+		Name:  "txlookuplimit",
+		Usage: "Number of recent blocks to index transactions-by-hash in (default = index all blocks)",
+		Value: 0,
```
Comment (karalabe) [inline cmd/utils/flags.go]: https://github.com/ethereum/go-ethereum/pull/20270/files#r346566579 :trollface: 

I think you were correct, please apply the diff

File: core/rawdb/accessors_chain.go [hunk 7]
```diff
@@ -290,6 +327,25 @@ func ReadBodyRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValue
 	return nil // Can't find the data anywhere.
 }
 
+// ReadCanonicalBodyRLP retrieves the block body (transactions and uncles) for the canonical
+// bloc at number, in RLP encoding.
```
Comment (karalabe) [inline core/rawdb/accessors_chain.go]: bloc->block

File: core/rawdb/accessors_chain.go [hunk 8]
```diff
@@ -290,6 +327,25 @@ func ReadBodyRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValue
...
+func ReadCanonicalBodyRLP(db ethdb.Reader, number uint64) rlp.RawValue {
+	// if it's an ancient one, we don't need the canonical hash
```
Comment (karalabe) [inline core/rawdb/accessors_chain.go]: Nit, please capitalize `If`

File: core/rawdb/accessors_indexes.go [hunk 9]
```diff
@@ -63,9 +63,30 @@ func WriteTxLookupEntries(db ethdb.KeyValueWriter, block *types.Block) {
 	}
 }
 
+// WriteTxLookupEntriesByHash is identical to WriteTxLookupEntries, but does not
+// require a full types.Block as input
+func WriteTxLookupEntriesByHash(db ethdb.KeyValueWriter, number []byte, hashes []common.Hash) {
```
Comment (karalabe) [inline core/rawdb/accessors_indexes.go]: Period after the comment.

Why is `number` a byte slice. That seems super unorthodox in an API.

File: core/rlp_test.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package core
```
Comment (karalabe) [inline core/rlp_test.go]: Please leave an empty line, otherwise godoc will push the entire copyright into the package doc.

File: core/rlp_test.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
+// Copyright 2019 The go-ethereum Authors
```
Comment (karalabe) [inline core/rlp_test.go]: This file seems super out of place. RLP iteration was already supposedly tested in package `rlp`. What does this add in addition? Also if it's just a sanity check that it works for blocks, shouldn't the entire thing be moved into `types`?

File: core/rawdb/accessors_indexes.go (hunk 9 above)
Comment (holiman) [reply core/rawdb/accessors_indexes.go]: I have no idea. Fixing

File: core/rlp_test.go (hunk 11 above)
Comment (holiman) [reply core/rlp_test.go]: So, why it doesn't sit in rlp is that it's testing that it works for `*types.Block`, as you guessed. And no, rlp iteration is new, it has no tests in the rlp package. 

I can move it to types...

Comment (holiman) [reply core/rlp_test.go]: Or rather, moving it to types is a bit difficult too, since I use GenerateChain and a lot of stuff from the core package...

**Comment by holiman** [comment]
Addressed most of the issues, (some open question above)

File: core/rlp_test.go (hunk 11 above)
Comment (holiman) [reply core/rlp_test.go]: See the comment on https://github.com/ethereum/go-ethereum/pull/20302/files#diff-16bfc2f96b958e952d2f3a0e32b8d38cR25 : 

```

// TestIterator tests some basic things about the ListIterator. A more
// comprehensive test can be found in core/rlp_test.go, where we can
// use both types and rlp without dependency cycles
```

**Comment by holiman** [comment]
Rebased to resolve conflicts

**Comment by holiman** [comment]
Due to the changes in https://github.com/ethereum/go-ethereum/pull/20695 , I had to rebase this. The only real conflict was in `sync.go`, where I moved this snippet: https://github.com/ethereum/go-ethereum/pull/20302/files#diff-75028a2cddf4c8ebf4133ad28d81ef7eR289 . 

The snippet is not originally my code, perhaps @rjl493456442 and @fjl can take a peek if it still looks correct?

**Comment by holiman** [comment]
Rebased

**Comment by karalabe** [comment]
I've played with this PR, seems to work as expected. Rebasing this will me messy to keep separate commits for Gary and Martin (mine can be fixuped into previous ones). If you guys want separate commits, sort it out :D, otherwise squashmerge.

**Comment by holiman** [comment]
I'm fine with a squashmerge

**Comment by holiman** [comment]
I was just looking into some travis failures..

**Comment by karalabe** [comment]
Oh shit, didn't see that the merge failed some tests, sorry.
